R/clean_variable_spelling.R

```diff
@@ -0,0 +1,152 @@
+#'   it is not recommended to use a dictionary without a grouping column 
+#'   specifying the columns in the data to work on.
+#'
+#' @return a data frame with re-defined data based on the dictionary 
+#' @seealso [clean_spelling()], which this function wraps.
+#' @author Zhian N. Kamvar
+#' @export
+#' @examples
+#' 
+#' # Set up wordlist ------------------------------------------------ 
+#' yesno <- c("Y", "N", "U", NA)
+#' dyesno <- c("Yes", "No", "Unknown", "missing")
+#' treatment_administered <- c(0:1, NA)
+#' dtreatment_administered <- c("Yes", "No", "missing")
+#' facility <- 1:10
+#' dfacility <- sprintf("Facility %s", format(1:10))
+#' age_group <- c(0, 10, 20, 30, 40, 50)
+#' dage_group <- c("0-9", "10-19", "20-29", "30-39", "40-49", "50+")
+#' 
+#' wordlist <- data.frame(
+#'   options = c(yesno, treatment_administered, facility, age_group),
+#'   values  = c(dyesno, dtreatment_administered, dfacility, dage_group),
+#'   orders  = c(1:4, 1:3, 1:10, 1:6),
+#'   stringsAsFactors = FALSE
+#' )
+#' wordlist$grp <- rep(c("readmission", "treatment_administered", "facility", "age_group"),
+#'                     c(4, 3, 10, 6))
+#'
+#' # Generate example data ------------------------------------------
+#' dat <- data.frame(
+#'   readmission = sample(yesno, 50, replace = TRUE),
+#'   treatment_administered = sample(treatment_administered, 50, replace = TRUE),
+#'   facility = sample(facility, 50, replace = TRUE),
+#'   age_group = sample(age_group, 50, replace = TRUE),
+#'   stringsAsFactors = FALSE
+#' )
+#'
+#' # Clean spelling based on wordlist ------------------------------ 
+#' wordlist # show the wordlist
+#' head(dat) # show the data
+#' 
+#' head(clean_variable_spelling(dat, wordlists = wordlist, group = "grp"))
+#' 
+#' # You can ensure the order of the factors are correct by specifying 
+#' # a column that defines order.
+#' dat[] <- lapply(dat, as.factor)
+#' as.list(head(dat))
+#' res <- clean_variable_spelling(dat, wordlists = wordlist, group = "grp", sort_by = "orders")
+#' head(res)
+#' as.list(head(res))
+
+clean_variable_spelling <- function(x = data.frame(), wordlists = list(), group = 3, sort_by = NULL, classes = NULL) {
+
+  if (length(x) == 0 || !is.data.frame(x)) {
+    stop("x must be a data frame")
+  }
+
+  if (is.null(classes)) {
+    classes <- i_find_classes(x)
+  }
+
+  # Define columns viable for manipulation ------------------------------------
+  # Because this is a global manipulator, only work on characters or factors
+  unprotected <- names(x)[classes %in% c("character", "factor")]
+
+  if (length(wordlists) == 0 || !is.list(wordlists)) {
+    stop("wordlists must be a list of data frames")
+  } 
+
+  # There is one big dictionary with groups -----------------------------------
+  if (is.data.frame(wordlists)) {
+
+    # There is a grouping column ----------------------------------------
+    if (!is.null(group) && length(group) == 1) {
```

Comment:
I may well be missing it here, so do tell me off, but to me it looks like providing `group` isn't a much needed flexibility. Given that people will probably design their dicts after our examples, we can impose this. We could impose that `wordlist` is either:

- a named list of 2-col `data.frame`s in which case names are matched against column names in the cleaned dataset

- a single `data.frame` with 2 or 3 columns, and if there is a 3rd column it is assumed to be the variable names

Note that for matching variable names against the input datasete, match should be optional as one may use a general-purpose dictionary; but I suspect that's what the current code does

Reply:
At any rate, I find `group` a bit misldeading a term. `variable_name`?

Reply:
I agree that `group` is misleading. I think I used it as a placeholder until a better name came up, but `variable_name` or `variables` or `spelling_variables` makes more sense.

I was designing this based on the dictionaries that I've seen for MSF in https://github.com/R4EPI/sitrep. Those dictionaries do not necessarily have the variable column as the third column.

Reply:
> Note that for matching variable names against the input datasete, match should be optional as one may use a general-purpose dictionary; but I suspect that's what the current code does

I'm not exactly sure what you mean here. Currently, there are two ways to go about this:

1. define the variables you want to work with
2. define a global dictionary

Regarding having a global dictionary, I think what you mentioned in https://github.com/reconhub/linelist/issues/36#issuecomment-465117954 is a good idea and will implement it.

Ultimately, what I want is for people to be able to use these wordlists for both spelling and defining label dictionaries with as little data manipulation on their side as possible.

Reply:
Implemented

Reply:
Wikid